### DISCUSSION TITLE
Periodic Uplink Lpp (Data Formats)

### DISCUSSION THREAD

**Discussion by meakash** [Lora-net/LoRaMac-node#1147] [Q&A] [answered]
There are three cases that I would like to confirm about the void PrepareTxFrame() in the periodic Uplink LPP for Nucleo L073 microcontroller.

**Case 1:** When the default firmware is used (as provided in the example i.e the Cayenne Lpp Uplink data) the data that is received is proper as intended. Which can be seen below on the ScreenShot of my Chirpstack application Server.
![image](https://user-images.githubusercontent.com/48630822/127816098-ea4313c0-bbae-49b9-b9ea-78903f26ccf8.png)

**Case 2:** When the C lines are changed (To send Hexadecimal Numbers as the Uplink) 
static void PrepareTxFrame( void )
{
    if( LmHandlerIsBusy( ) == true )
    {
        return;
    }
printf("Sending Data");
uint32_t i = 0;
AppData.Buffer[i++] = 0xFE;
AppData.Buffer[i++] = 0x4F;
AppData.Buffer[i++] = 0x63;
AppData.Buffer[i++] = 0x00;
AppData.BufferSize = i;
    if( LmHandlerSend( &AppData, LmHandlerParams.IsTxConfirmed ) == LORAMAC_HANDLER_SUCCESS )
    {
        // Switch LED 1 ON
        GpioWrite( &Led1, 1 );
        TimerStart( &Led1Timer );
    }
}
The following is the output received as seen in the screenshot below:
![image](https://user-images.githubusercontent.com/48630822/127817087-5d3190ff-a82f-4095-999e-f5e16691d67e.png)

As can be seen the data field is Empty, but when I check the LoRaWAN Frames there is something that I receive in the CID Field which corresponds to the hexadecimal Numbers I have programmed to be sent and can be verified in the screenshot below: 
![image](https://user-images.githubusercontent.com/48630822/127817318-b31d4532-b46d-473a-8588-4c9f4349429e.png)

**Case 3:** When the C lines are changed (To send String as the Uplink)
static void PrepareTxFrame( void )
{
    if( LmHandlerIsBusy( ) == true )
    {
        return;
    }
/*without cayennelpp*/
/*To Send a Char Array*/
char data[]="Hello World";
sprintf((char*)AppData.Buffer, "%s", data);
AppData.BufferSize = strlen ((char*)AppData.Buffer);
    if( LmHandlerSend( &AppData, LmHandlerParams.IsTxConfirmed ) == LORAMAC_HANDLER_SUCCESS )
    {
        // Switch LED 1 ON
        GpioWrite( &Led1, 1 );
        TimerStart( &Led1Timer );
    }
}
![image](https://user-images.githubusercontent.com/48630822/127818487-74b97190-621b-4d2c-87f5-d8e467e05376.png)
Still there is null in the payload Filed and then Hello world is received in the CID field of LoRaWAN Frames Part.

**Please help me out by telling what am i missing so that I am able to send Numbers as well as strings as the uplink data that can be received in the Payload Field.** As per requirement i cant proceed with Cayenne LPP. Both this method of sending Numbers and Strings work with the ST I-Cube LRWAN (ST's modified stack for Nucleo L LoRaWAN devices). Due to requirement, I have to use LoRaMAC 4.5.2, so any help will be appreciated.

**Comment by Pirakutu** [comment, marked as answer]
You are missing the port, port 0 is  no valid payload port, i think that is the problem

**Reply by meakash** [reply]
@Pirakutu  Isn't the port selected by the following line: 

#define LORAWAN_APP_PORT                            2 
this is in main.c

**Reply by Pirakutu** [reply]
Yeah but it needs to be loaded into the AppData structure, which is then passed to the stack with the LmHandlerSend function. I think it has to be something in this direction, cuz you can see that the other ULs come on port 0 and the first one comes on port 2, as defined with the LORAWAN_APP_PORT define.

**Reply by meakash** [reply]
Well Thank you so much for pointing this out. thanks a lot!!!
